kbmd/piv.c

```diff
@@ -0,0 +1,1229 @@
+/*
+ * This file and its contents are supplied under the terms of the
+ * Common Development and Distribution License ("CDDL"), version 1.0.
+ * You may only use this file in accordance with the terms of version
+ * 1.0 of the CDDL.
+ *
+ * A full copy of the text of the CDDL should have accompanied this
+ * source.  A copy of the CDDL is also available via the Internet at
+ * http://www.illumos.org/license/CDDL.
+ */
+
+/*
+ * Copyright 2020 Joyent, Inc.
+ */
+
+#include <sys/sysmacros.h>
+#include <openssl/err.h>
+#include <openssl/x509.h>
+#include <openssl/x509v3.h>
+#include "kbmd.h"
+#include "pivy/libssh/sshbuf.h"
+#include "pivy/libssh/ssherr.h"
+#include "pivy/libssh/sshkey.h"
+#include "pivy/tlv.h"
+
+/*  We need the piv_cert_comp enum */
+#include "pivy/piv-internal.h"
+
+#define funcerrf(cause, fmt, ...)       \
+    errf(__func__, cause, fmt , ##__VA_ARGS__)
+
+#define	ADMIN_KEY_LENGTH 24
+
+/*
+ * How long after initializing a PIV token we wait for the initialized PIV
+ * token to appear (in ms).
+ */
+uint_t init_timeout = 2000;
+uint_t init_wait = 500;
+
+kbmd_token_t *sys_piv;
+
+/*
+ * The default values of a piv token that hasn't been setup.
+ */
+static const uint8_t DEFAULT_ADMIN_KEY[ADMIN_KEY_LENGTH] = {
+	0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
+	0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
+	0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
+};
+static const char DEFAULT_PIN[] = "123456";
+static const char DEFAULT_PUK[] = "12345678";
+
+static const char *piv_pin_str(enum piv_pin);
+
```

Comment:
Not sure, but is a small paragraph in order here mentioning that all functions with sensitive local variables MUST explicit_bzero() before returning?

Reply:
As I mentioned in another comment, we explicitly use allocators for custr_t and nvlist_t that do an explicit_bzero on free. For things not using that, it'd be true for things not just in this file (and should be doing that already where possible -- i.e. for some openssl calls, we have no choice but to rely on them to do that where they do their own free methods for things).